pkg/generators/markers.go

```diff
@@ -0,0 +1,225 @@
+	//        },
+	//      },
+	//    },
+	//  }
+	//
+	markers, err := parseMarkers(comments, markerPrefix)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("failed to parse marker comments: %w", err)
+	}
+	nested, err := nestMarkers(markers)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("invalid marker comments: %w", err)
+	}
+
+	// Parse the map into a CommentTags type by marshalling and unmarshalling
+	// as JSON in leiu of an unstructured converter.
+	out, err := json.Marshal(nested)
+	if err != nil {
+		return CommentTags{}, fmt.Errorf("failed to marshal marker comments: %w", err)
+	}
+
+	var commentTags CommentTags
+	if err = json.Unmarshal(out, &commentTags); err != nil {
+		return CommentTags{}, fmt.Errorf("failed to unmarshal marker comments: %w", err)
+	}
+
+	return commentTags, nil
+}
+
+func parseMarkers(markerComments []string, prefix string) (map[string]any, error) {
+	markers := types.ExtractCommentTags("+", markerComments)
+
+	// Parse the values as JSON
+	result := map[string]any{}
+	for key, value := range markers {
+		if !strings.HasPrefix(key, prefix) {
+			// we only care about validation markers for now
+			continue
+		}
+
+		newKey := strings.TrimPrefix(key, prefix)
+
+		// Skip ref markers
+		if len(value) == 1 && regexp.MustCompile(`^(?:\s*ref\s{0,1}\().*\)\s*`).MatchString(value[0]) {
+			continue
+		}
+		if len(newKey) == 0 {
+			return nil, fmt.Errorf("cannot have empty key for marker comment")
+		} else if len(value) == 0 || (len(value) == 1 && len(value[0]) == 0) {
+			// Empty value means key is implicitly a bool
+			result[newKey] = true
+			continue
+		}
+
+		newVal := []any{}
+		for _, v := range value {
+			var unmarshalled interface{}
+			err := json.Unmarshal([]byte(v), &unmarshalled)
+			if err != nil {
+				return nil, fmt.Errorf("invalid value for key %v: %w", key, err)
+			}
+
+			newVal = append(newVal, unmarshalled)
+		}
+
+		if len(newVal) == 1 {
+			result[newKey] = newVal[0]
+		} else {
+			result[newKey] = newVal
+		}
+	}
+	return result, nil
+}
+
+// Converts a map of:
+//
+//	"a:b:c": 1
+//	"a:b:d": 2
+//	"a:e": 3
+//	"f": 4
+//
+// Into:
+//
+//	 map[string]any{
+//	   "a": map[string]any{
+//		      "b": map[string]any{
+//		          "c": 1,
+//				  "d": 2,
+//			   },
+//			   "e": 3,
+//		  },
+//		  "f": 4,
+//	 }
+func nestMarkers(markers map[string]any) (map[string]any, error) {
+	nested := make(map[string]any)
+	var errs []error
+	for key, value := range markers {
+		var err error
+		keys := strings.Split(key, ":")
+		nested, err = putNestedValue(nested, keys, value)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
+	}
+
+	return nested, nil
+}
+
+// Recursively puts a value into the given keypath, creating intermediate maps
+// and slices as needed. If a key is of the form `foo[bar]`, then bar will be
+// treated as an index into the array foo. If bar is not a valid integer, then
+// the key will be ignored.
```

Comment:
Could we make it return an error instead of ignoring non-int array subscripts?

Reply:
I think the logic does already return an error, the comment just needs to be updated.